### PR TITLE
add comment on acb_zio_dummy

### PR DESCRIPTION
**Pull request by problame**
Thanks to George Wilson for clarifying this on Slack.

Signed-off-by: Christian Schwarz <christian.schwarz@nutanix.com>

### CODE REVIEW

File: module/zfs/arc.c [hunk 1]
```diff
@@ -6051,6 +6068,10 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 				acb->acb_noauth = noauth_read;
 				acb->acb_nobuf = no_buf;
 				acb->acb_zb = *zb;
+				/*
+				 * XXX couldn't there be a arc_read with
+				 * done == NULL and pio != NULL?
+				 */
```
Comment (problame) [inline module/zfs/arc.c:6074]: @grwilson  please comment on this XXX?

File: module/zfs/arc.c [hunk 2]
```diff
@@ -6032,6 +6032,23 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
...
+			/*
+			 * If there are multiple threads reading the same block
+			 * and that block is not yet in the ARC, then only one
+			 * thread will do the physical I/O and all other
+			 * threads will wait until that I/O completes.
+			 * Synchronous reads use the b_cv whereas nowait reads
+			 * register a callback. Both are signalled/called in
+			 * arc_read_done.
+			 *
+			 * Errors of the physical I/O need to be propagated to
```
Comment (grwilson) [inline module/zfs/arc.c:6044]: NIT: `Errors of the logical I/O may need...` The `zio_flags` may indicate that errors should not be propagated.

File: module/zfs/arc.c (hunk 1 above)
Comment (grwilson) [reply module/zfs/arc.c:6074]: Passing in a `pio` typically means that the caller wants some kind of interlocking of the I/Os that are issued and could be interested in propagating errors up the zio tree. Passing in a `done` func means that the caller wants to do something with the data.  So what would it mean to pass in a `pio` but a NULL `done`? Maybe a prefetch that cares about a failed I/O and wants to do something about it? For completeness, this seems like an overlooked case that could be solved by unconditionally creating the `acb`. Another option would be to add an `IMPLY(done == NULL, pio == NULL)` or something check like that. This way any code that were to add such a call would ensure the proper testing of any changes (I don't think we currently have a caller which passes this condition). If you do want to add a check, then it should also be added in the regular cache miss code path as well.

**Comment by problame** [comment]
Thanks for the review, @grwilson. Addressed your nit.
Will open a separate PR to add `IMPLY(done == NULL, pio == NULL)` . 
Please approve the change if it looks good to you.

File: module/zfs/arc.c [hunk 3]
```diff
@@ -6032,6 +6032,23 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
...
+			 * and that block is not yet in the ARC, then only one
+			 * thread will do the physical I/O and all other
+			 * threads will wait until that I/O completes.
+			 * Synchronous reads use the b_cv whereas nowait reads
+			 * register a callback. Both are signalled/called in
+			 * arc_read_done.
+			 *
+			 * Errors of the physical I/O may need to be propagated
+			 * to the pio. For synchronous reads, we simply restart
+			 * this function and it will re-asses.  Nowait reads
```
Comment (behlendorf) [inline module/zfs/arc.c:6046]:
```suggestion
			 * this function and it will reassess.  Nowait reads
```

**Review by behlendorf — APPROVED** [review]

**Review by grwilson — APPROVED** [review]

**Pull request merged**